polkadot/runtime/rococo/src/lib.rs

```diff
@@ -1799,6 +1804,54 @@ sp_api::impl_runtime_apis! {
 		}
 	}
 
+	impl xcm_fee_payment_runtime_api::XcmPaymentApi<Block, RuntimeCall> for Runtime {
+		fn query_acceptable_payment_assets(xcm_version: xcm::Version) -> Result<Vec<VersionedAssetId>, XcmPaymentApiError> {
+			if !matches!(xcm_version, 3 | 4) {
+				return Err(XcmPaymentApiError::UnhandledXcmVersion);
+			}
+			Ok([VersionedAssetId::V4(xcm_config::TokenLocation::get().into())]
+				.into_iter()
+				.filter_map(|asset| asset.into_version(xcm_version).ok())
+				.collect())
+		}
+
+		fn query_weight_to_asset_fee(weight: Weight, asset: VersionedAssetId) -> Result<u128, XcmPaymentApiError> {
+			let local_asset = VersionedAssetId::V4(xcm_config::TokenLocation::get().into());
+			let asset = asset
+				.into_version(4)
+				.map_err(|_| XcmPaymentApiError::VersionedConversionFailed)?;
+
+			if  asset != local_asset { return Err(XcmPaymentApiError::AssetNotFound); }
+
+			Ok(WeightToFee::weight_to_fee(&weight))
+		}
+
+		fn query_xcm_weight(message: VersionedXcm<RuntimeCall>) -> Result<Weight, XcmPaymentApiError> {
+			let mut message = message
+				.try_into()
+				.map_err(|_| XcmPaymentApiError::VersionedConversionFailed)?;
+			<xcm_config::XcmConfig as xcm_executor::Config>::Weigher::weight(&mut message)
+				.map_err(|error| {
+					log::error!("Error when querying XCM weight: {:?}", error);
+					XcmPaymentApiError::WeightNotComputable
+				})
+		}
+
+		fn query_delivery_fees(destination: VersionedLocation, message: VersionedXcm<()>) -> Result<VersionedAssets, XcmPaymentApiError> {
+			let destination = destination
+				.try_into()
+				.map_err(|_| XcmPaymentApiError::VersionedConversionFailed)?;
+			let message = message
+				.try_into()
+				.map_err(|_| XcmPaymentApiError::VersionedConversionFailed)?;
+			let (_, fees) = xcm_config::XcmRouter::validate(&mut Some(destination), &mut Some(message)).map_err(|error| {
+				log::error!("Error when querying delivery fees: {:?}", error);
+				XcmPaymentApiError::Unroutable
+			})?;
+			Ok(VersionedAssets::from(fees))
```

Comment:
A client might not understand the returned XCM version. The returned `VersionedAssets` should follow the version of one of the inputs. For instance, if the `destination` and the `message` are both v3, the returned assets should also be in the v3 form.
The current code will always return the form of the latest XCM version, which might break some clients (no one uses this API at the moment, but IMHO, we need a proper way of handling this for future XCM versions).

For the same reason, we have the `xcm_version` parameter in the `query_acceptable_payment_assets` method.

Another question, though: what if the `destination` and the `message` are expressed via different XCM versions? What version should the returned assets follow?

The same goes for Westend.

Reply:
> The returned VersionedAssets should follow the version of one of the inputs. For instance, if the destination and the message are both v3, the returned assets should also be in the v3 form.

👍 

> Another question, though: what if the destination and the message are expressed via different XCM versions? What version should the returned assets follow?

use the newest version possible

Reply:
Done. The implementation is located in pallet-xcm, as per the comment below.
I added the `IdentifyVersion` implementation for `VersionedXcm` to implement the "use the newest version between them" logic